src/metabase/driver/csv.clj

```diff
@@ -0,0 +1,242 @@
+    :day-of-week     (hsql/call :day_of_week expr)
+    :day-of-month    (hsql/call :day_of_month expr)
+    :day-of-year     (hsql/call :day_of_year expr)
+    :week            (trunc-with-format "YYYYww" expr) ; Y = week year; w = week in year
+    :week-of-year    (hx/week expr)
+    :month           (trunc-with-format "yyyyMM" expr)
+    :month-of-year   (hx/month expr)
+    ;; Rounding dates to quarters is a bit involved but still doable. Here's the plan:
+    ;; *  extract the year and quarter from the date;
+    ;; *  convert the quarter (1 - 4) to the corresponding starting month (1, 4, 7, or 10).
+    ;;    (do this by multiplying by 3, giving us [3 6 9 12]. Then subtract 2 to get [1 4 7 10])
+    ;; *  Concatenate the year and quarter start month together to create a yyyyMM date string;
+    ;; *  Parse the string as a date. :sunglasses:
+    ;;
+    ;; Postgres DATE_TRUNC('quarter', x)
+    ;; becomes  PARSEDATETIME(CONCAT(YEAR(x), ((QUARTER(x) * 3) - 2)), 'yyyyMM')
+    :quarter         (parse-datetime "yyyyMM"
+                                     (hx/concat (hx/year expr) (hx/- (hx/* (hx/quarter expr)
+                                                                           3)
+                                                                     2)))
+    :quarter-of-year (hx/quarter expr)
+    :year            (hx/year expr)))
+
+;; TODO - maybe rename this relative-date ?
+(defn- date-interval [unit amount]
+  (if (= unit :quarter)
+    (recur :month (hx/* amount 3))
+    (hsql/call :dateadd (hx/literal unit) amount :%now)))
+
+
+(defn- humanize-connection-error-message [message]
+  (condp re-matches message
+    #"^A file path that is implicitly relative to the current working directory is not allowed in the database URL .*$"
+    (driver/connection-error-messages :cannot-connect-check-host-and-port)
+
+    #"^Database .* not found .*$"
+    (driver/connection-error-messages :cannot-connect-check-host-and-port)
+
+    #"^Wrong user name or password .*$"
+    (driver/connection-error-messages :username-or-password-incorrect)
+
+    #".*" ; default
+    message))
+
+(defn- string-length-fn [field-key]
+  (hsql/call :length field-key))
+
+(def ^:private date-format-str   "yyyy-MM-dd HH:mm:ss.SSS zzz")
+(def ^:private h2-date-formatters (driver/create-db-time-formatters date-format-str))
+(def ^:private h2-db-time-query  (format "select formatdatetime(current_timestamp(),'%s') AS VARCHAR" date-format-str))
+
+(defrecord CSVDriver []
+  clojure.lang.Named
+  (getName [_] "CSV"))
+
+(u/strict-extend CSVDriver
+  driver/IDriver
+  (merge (sql/IDriverSQLDefaultsMixin)
+         {:date-interval                     (u/drop-first-arg date-interval)
+          :details-fields                    (constantly [{:name         "db"
+                                                           :display-name "Connection String"
+                                                           :placeholder  "file:/Users/camsaul/bird_sightings/toucans"
```

Comment:
excellent placeholder 💯

Reply:
Should work on your machine ;) 
Added a better one.

Reply:
Oh I wasn’t being sarcastic I genuinely liked it